src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,219 @@
+            var caretLocation = new LinePosition(originalOpeningLinePosition.Line, originalOpeningLinePosition.Character + 1);
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation);
+        }
+
+        public virtual Task<BraceCompletionResult?> GetTextChangesAfterCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual Task<BraceCompletionResult?> GetTextChangeAfterReturnAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+            => SpecializedTasks.Default<BraceCompletionResult?>();
+
+        public virtual async Task<bool> CanProvideBraceCompletionAsync(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+        {
+            if (OpeningBrace != brace)
+            {
+                return false;
+            }
+
+            // check that the user is not typing in a string literal or comment
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
+        }
+
+        public async Task<BraceCompletionContext?> GetCompletedBraceContextAsync(Document document, int caretLocation, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var leftToken = root.FindTokenOnLeftOfPosition(caretLocation);
+            var rightToken = root.FindTokenOnRightOfPosition(caretLocation);
+
+            if (IsValidOpeningBraceToken(leftToken) && IsValidClosingBraceToken(rightToken))
+            {
+                return new BraceCompletionContext(document, leftToken.GetLocation().SourceSpan.Start, rightToken.GetLocation().SourceSpan.End, caretLocation);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks if the already inserted token is a valid opening token at the position in the document.
+        /// By default checks that the opening token is a valid token at the position and not in skipped token trivia.
+        /// </summary>
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            if (token.SpanStart != position)
+            {
+                return SpecializedTasks.False;
+            }
+
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            return Task.FromResult(IsValidOpeningBraceToken(token) && !ParentIsSkippedTokensTrivia(syntaxFactsService, token));
```

Comment:
aside: is it possible for any of these virtual methods to become sync?  or is there always at elast one case where we do something async in one of hte sublasses?

Reply:
There's a few cases in the implementations where we do async work :(